### ISSUE TITLE
Implement a more advanced way to ignore/suppress issues

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#4068]
Detekt should have an advanced way to exclude/ignore/suppress issues by confiugration. And it should be generic so it could be applied to any rule. Right now we just allow to ignore files completely.

## Context
Right now we have this kind of configurations build-in the rules that clearly need this. Examples:

- LongMethod.ignoreAnnotated
- LongParameterList.ignoreAnnotated
- ForbiddenPublicDataClass.ignorePackages
- ForbiddenVoid.ignoreOverridden
- MagicNumber.ignoreEnums
- ReturnCount.excludedFunctions
- ...

Detekt is getting more and more users so the use cases are growing and we get more and more request to adapt our rules for these use cases, Examples: #4062, #4061 and #2231 (there are lots more)

I don't know exactly how this api should look like but I think that detekt really need this feature.

So, any ideas about how to implement this or how the api should be?

**Comment by cortinico** [comment]
I thought a bit about this issue. I believe a generic `ignoreAnnotated` is not that hard to implement. 

Whenever we report a finding we need to make sure we report the `ktElement` correctly:

https://github.com/detekt/detekt/blob/198c53cdcd61c6004911352bb1855917420e23c1/detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Entity.kt#L14-L19

On that element we can easily traverse the AST and check if him or any of the parent is annotated with any of the `ignoreAnnotated` annotation.

I believe annotations exceptions are the more critical to address right now, with the advent of Compose. I'm unsure about the other exclusion mechanism and can't think at anything really generic that would be worth to implement (happy to hear the community opinion on this).

**Comment by BraisGabin** [comment]
I can imagine `ignoreInFunction` `IgnoreInClass`. Same as `ignoreAnnotated`. It traverse the AST and check if it is inside a function with that name or a class with that name. There are rules where some of them will not applied, but that should not be a problem. If you add `ignoreInFunction` in `LongClass` it will do nothing because the class (usually) is not inside a function.

**Comment by cortinico** [comment]
> The required change is to incorporate the rule configuration into the traversal.

Agree.
The point here is deciding what we want to inspect for (annotations, packages, etc.)

**Comment by BraisGabin** [comment]
I'm looking at this. I want to implement the first `Suppressor`. It will allow to suppress findings with annotations. With this we will have the base code so any new `Suppressor` should be easier to add. Right now detekt will not allow custom `Suppressor`s but that could be archived in later versions. For example, my current proposal for the core of 2.0 allow this already.

But, I found that right now if someone *really* need this functionality it can be archived with a Custom `ReportingExtension`. It's not very combinient for this use case but it's doable.

**Comment by BraisGabin** [comment]
I talked about adding one more of this here: https://github.com/detekt/detekt/issues/4080#issuecomment-932754009 and I just implemented one more here: #4148

My idea here is to experiment with this new idea of suppressors so we can see some examples and then create a "standard". 

For the documentation part I think that once we make them "standard" we should auto generate it to avoid getting it unsync. meanwhile, should we add something in our documentation or just add it in the changelog?
